File: Template/Naked Objects Server/Template.Model/Student.cs
using NakedObjects;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using NakedFramework;

namespace Template.Model
{
    public class Student
    {
        #region Injected Services

        public TeacherRepository Teachers { set; protected get; }

        public IDomainObjectContainer Container { set; protected get; }

        #endregion

        [Hidden]
        public virtual int Id { get; set; }

        [MemberOrder(1), Title]
        public virtual string FullName { get; set; }

        [MemberOrder(2), Range(9, 13)]
        public virtual int CurrentYearGroup { get; set; }

        [MemberOrder(4), Optionally,]
        public virtual Teacher PersonalTutor { get; set; }

        public IList<Teacher> ChoicesPersonalTutor() => Teachers.AllTeachers().ToList();


        [MemberOrder(3), Eagerly(Do.Rendering)]
        [TableView(false, "Subject", "SetName", "Teacher")]
        public virtual ICollection<Set> Sets { get; set; } = new List<Set>();

        public IQueryable<SubjectReport> ListRecentReports([Optionally] Subject forSubject)
        {
            int id = this.Id;
            var studentReps = Container.Instances<SubjectReport>().Where(sr => sr.Student.Id == id);
            if (forSubject != null)
            {
                var subjId = forSubject.Id;
                studentReps = studentReps.Where(sr => sr.Subject.Id == subjId);
            }
            return studentReps.OrderByDescending(sr => sr.Date);
        }

        public SubjectReport CreateNewReport(Subject sub)
        {
            var rep = Container.NewTransientInstance<SubjectReport>();
            rep.Student = this;
            rep.Subject = sub;
            return rep;
        }

        public void SendMessage(string subject, string message) => Container.InformUser("Message sent to " + FullName);

    }
}

File: Template/NOF2 Server/Model.Template/Student.vb

Imports System
Imports System.Collections.Generic
Imports System.ComponentModel.DataAnnotations

Imports NakedFramework


    Public Class Student
        Public Property Teachers As TeacherRepository
        Public Property Container As IContainer

    Public Overridable Property Id As Integer
    <DemoProperty(Order:=1)>
    Public Overridable Property FullName As String
    <DemoProperty(Order:=2), Range(9, 13)>
    Public Overridable Property CurrentYearGroup As Integer
    <DemoProperty(Order:=4)>
    Public Overridable Property PersonalTutor As Teacher

    Public Function ChoicesPersonalTutor() As IList(Of Teacher)
        Return Teachers.AllTeachers().ToList()
    End Function

    <DemoProperty(Order:=3)>
    <TableView(False, "Subject", "SetName", "Teacher")>
        Public Overridable Property Sets As ICollection(Of [Set]) = New List(Of [Set])()

    Public Function ListRecentReports(
 ByVal forSubject As Subject) As IQueryable(Of SubjectReport)
        Dim id As Integer = Me.Id
        Dim studentReps = Container.AllInstances(Of SubjectReport)().Where(Function(sr) sr.Student.Id = id)

        If forSubject IsNot Nothing Then
            Dim subjId = forSubject.Id
            studentReps = studentReps.Where(Function(sr) sr.Subject.Id = subjId)
        End If

        Return studentReps.OrderByDescending(Function(sr) sr.Date)
    End Function

    Public Function CreateNewReport(ByVal [sub] As Subject) As SubjectReport
        Dim rep = Container.CreateTransientInstance(Of SubjectReport)()
        rep.Student = Me
            rep.Subject = [sub]
            Return rep
        End Function

        Public Sub SendMessage(ByVal subject As String, ByVal message As String)
        Container.AddMessageToBroker("Message sent to " & FullName)
    End Sub

        Public Overrides Function ToString() As String
            Return FullName
        End Function
    End Class

